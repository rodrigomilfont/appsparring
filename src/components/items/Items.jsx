import React from 'react';
import { Link } from 'react-router-dom';

import axios from 'axios';
import Header from '../header/Header';

class Items extends React.Component {
  constructor(props) {
    super(props);
    console.log('props: ', props);

    this.state = {
      termHeader: '',
    };
    this.handleHeader = searchTerm => {
      this.setState({ termHeader: searchTerm });
      this.makeRequest(searchTerm);
    };

    this.makeRequest = searchTerm => {
      axios
        .get(`https://api.mercadolibre.com/sites/MLA/search?q=${searchTerm}`)
        .then(response => {
          this.setState({ resultSearch: response.data });
        });
    };
  }
  componentDidMount() {}
  render() {
    return (
      <div className="grid">
        <Header onSearchTermChange={this.handleHeader} />
        <main className="content">
          <article>
            <Link to="/">Home</Link>
            <h1>Search Term : {this.state.termHeader}</h1>
            <pre>
              <code>{JSON.stringify(this.state.resultSearch, null, 4)}</code>
            </pre>
          </article>
        </main>
      </div>
    );
  }
}

export default Items;
