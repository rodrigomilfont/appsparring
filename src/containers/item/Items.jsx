import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import Header from '../../components/header/Header';

const Items = props => {
  const {
    searchTerm,
    searchs,
    isFetching,
    lastUpdate,
    error,
    errorMsg,
  } = props;
  return (
    <div className="grid">
      <Header />
      <main className="content">
        <article>
          <Link to="/">Home</Link>
          <br />
          {/* // Super loading */}
          {isFetching && searchs.length === 0 && !error && <h2>Loading...</h2>}
          {error && <h2 className="error" >Search failed...try again! Msg: {errorMsg} </h2>}
          {lastUpdate && (
            <span>
              Last updated at {new Date(lastUpdate).toLocaleTimeString()}
            </span>
          )}
          {searchs.length !== 0 && (
            <div>
              <h2>Search Term Items : {searchTerm}</h2>
              <pre>
                <code>{JSON.stringify(searchs, null, 4)}</code>
              </pre>
            </div>
          )}
        </article>
      </main>
    </div>
  );
};

Items.propTypes = {
  searchs: PropTypes.arrayOf(PropTypes.object),
  isFetching: PropTypes.bool,
  error: PropTypes.bool,
  errorMsg: PropTypes.string,
  lastUpdate: PropTypes.number,
  searchTerm: PropTypes.string,
};

Items.defaultProps = {
  searchs: [],
  isFetching: false,
  error: false,
  errorMsg: '',
  lastUpdate: null,
  searchTerm: '',
};

const mapStateToProps = state => {
  const { searchTerm, resultBySearchTerm, failedSearch } = state;

  const { error, errorMsg } = failedSearch || { error: false, errorMsg: '' };

  const { isFetching, lastUpdate, items: searchs } = resultBySearchTerm[
    searchTerm
  ] || {
    isFetching: false,
    items: [],
  };

  return {
    searchTerm,
    isFetching,
    lastUpdate,
    searchs,
    error,
    errorMsg,
  };
};

export const Unwrapped = Items

export default connect(mapStateToProps)(Items);
