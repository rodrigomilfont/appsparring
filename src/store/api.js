export { fetchSearch } from './search/endpoint';

export function checkStatus(response) {
  console.log('response: ', response);

  console.log('status: ', response.status);

  return response;
}
