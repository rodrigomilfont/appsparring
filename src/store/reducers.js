import { combineReducers } from 'redux';
import search from './search/reducers';

export default combineReducers({
  search,
});
