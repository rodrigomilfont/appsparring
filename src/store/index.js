import { createStore, compose, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from './reducers';
import * as api from './api';

const store = createStore(
  rootReducer,
  compose(
    applyMiddleware(thunk.withExtraArgument(api)),
    typeof window === 'object' &&
    typeof window.devToolsExtension !== 'undefined'
      ? window.devToolsExtension()
      : f => f,
  ),
);

export default store;
