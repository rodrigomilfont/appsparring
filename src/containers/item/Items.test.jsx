import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import store from '../../store';
import {
  setSearchTerm,
  failedSearch,
  requestSearch,
} from '../../actionCreators';
import Items, {
  Unwrapped as UnwrappedItems,
} from '../../containers/item/Items';

Enzyme.configure({ adapter: new Adapter() });

function setup() {
  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <Items />
      </MemoryRouter>
    </Provider>,
  );
  return {
    wrapper,
  };
}

describe('Unwrapped', () => {
  it('renders without crashing', () => {
    const wrapper = shallow(<UnwrappedItems />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Connected Items', () => {
  const searchTerm = 'Goku';

  it('change search term', () => {
    store.dispatch(setSearchTerm(searchTerm));

    const { wrapper } = setup();
    expect(wrapper.find('input').props().value).toEqual(searchTerm);
  });

  it('failedSearch test', () => {
    store.dispatch(failedSearch(searchTerm, 'ERROR'));

    const { wrapper } = setup();
    expect(wrapper.find('.error').length).toEqual(1);
  });

  // #bug mount enzyme https://github.com/airbnb/enzyme/issues/1229
  // xit('requestSearch test', () => {
  //   store.dispatch(requestSearch(searchTerm));
  //   const { wrapper } = setup();
  //
  //   expect(wrapper.find('.loading').length).toEqual(1);
  // });
});
