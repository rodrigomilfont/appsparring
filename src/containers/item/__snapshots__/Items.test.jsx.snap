// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unwrapped renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Items
    error={false}
    errorMsg=""
    isFetching={false}
    lastUpdate={null}
    searchTerm=""
    searchs={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Header
          error={false}
          errorMsg=""
          isFetching={false}
          lastUpdate={null}
          searchTerm=""
          searchs={Array []}
          showSearch={true}
        />,
        <main
          className="content"
        >
          <article>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
            <br />
          </article>
        </main>,
      ],
      "className": "grid",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "error": false,
          "errorMsg": "",
          "isFetching": false,
          "lastUpdate": null,
          "searchTerm": "",
          "searchs": Array [],
          "showSearch": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <article>
            <Link
              replace={false}
              to="/"
            >
              Home
            </Link>
            <br />
          </article>,
          "className": "content",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                replace={false}
                to="/"
              >
                Home
              </Link>,
              <br />,
              false,
              false,
              null,
              false,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": "Home",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": "Home",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            false,
            false,
            null,
            false,
          ],
          "type": "article",
        },
        "type": "main",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Header
            error={false}
            errorMsg=""
            isFetching={false}
            lastUpdate={null}
            searchTerm=""
            searchs={Array []}
            showSearch={true}
          />,
          <main
            className="content"
          >
            <article>
              <Link
                replace={false}
                to="/"
              >
                Home
              </Link>
              <br />
            </article>
          </main>,
        ],
        "className": "grid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "error": false,
            "errorMsg": "",
            "isFetching": false,
            "lastUpdate": null,
            "searchTerm": "",
            "searchs": Array [],
            "showSearch": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <article>
              <Link
                replace={false}
                to="/"
              >
                Home
              </Link>
              <br />
            </article>,
            "className": "content",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Link
                  replace={false}
                  to="/"
                >
                  Home
                </Link>,
                <br />,
                false,
                false,
                null,
                false,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Home",
                  "replace": false,
                  "to": "/",
                },
                "ref": null,
                "rendered": "Home",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              false,
              false,
              null,
              false,
            ],
            "type": "article",
          },
          "type": "main",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
